### PR TITLE
Fix Set Header Values & Consolidate SeleniumExtentions Wait Methods

### PR DESCRIPTION
**Pull request by AngelRodriguez8008**
### Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Other (updates to documentation, formatting, etc.)

### Description
When a tests is trying to set or get a field value, it was ignored in which context the field was. Header or Entity the container was the same. Repeated fields, in fields & form body, are not setting correctly. Also can be a problem in BPF, Quick Forms etc. 
With new this approach field controls will be search in the desired context.
See bug #719 .

SeleniumExtentions (Wait Methods) are consolidated same signatures & logic for WaitUntil (Available, Visible, Clickable). 
Fix also bug introduced in #707  see [comment](https://github.com/microsoft/EasyRepro/pull/707#discussion_r368233464)

### Issues addressed
#719 
#707 

### All submissions:

- [x] My code follows the code style of this project.
- [x] Do existing samples that are effected by this change still run?
- [x] I have added samples for new functionality. 
- [x] I raise detailed error messages when possible.
- [x] My code does not rely on labels that have the option to be hidden.

### Which browsers was this tested on?

- [x] Chrome
- [ ] Firefox
- [ ] IE
- [ ] Edge

### CODE REVIEW

**Comment by KhoaAN** [comment]
@AngelRodriguez8008 ,
Something broke the UCI submenu navigation, I tried to fix it but its not easy so i'll leave it to you. Description of the issue:
1. When you reproduce this issue, please update the AppElementReference for these guys like below (Microsoft must have updated the UCI Nav control so the Ids no longer matched):
`            //Navigation
            { "Nav_AreaButton"       , "//button[contains(@data-lp-id,'sitemap-areaSwitcher-expand-btn')]"},
            { "Nav_AreaMenu"       , "//*[@data-lp-id=\"sitemap-area-switcher-flyout\"]"},`
2. It failed with error message "{"The given key was not present in the dictionary."}" from the code locations below: 
image

image

My Menu nav setup:
image

**Comment by AngelRodriguez8008** [comment]
Hi @KhoaAN,

I found the bug, fixed it & add two tests for this sub-menu. What you mention is true the AppElementReference is not fully updated.

How did you find this error? I couldn't find any associated test.

Thanks for the review. 
br

**Comment by KhoaAN** [comment]
> Hi @KhoaAN,
> 
> I found the bug, fixed it & add two tests for this sub-menu. What you mention is true the AppElementReference is not fully updated.
> 
> How did you find this error? I couldn't find any associated test.
> 
> Thanks for the review.
> br

Hi @AngelRodriguez8008 ,
Thanks for applying the fix. It stopped me from checking the actual functionality of setting Lookup field on form and header. I'll get back to those. 
I debug using a real D365 application that are being built and its using the latest UCI version. The AppElementReference had worked fine until last week or early this week. It must be Microsoft pushing out D365 changes and one of them might be to do with changing that Nav Control causing the generated data* id not matching our definition in AppElementReference.

**Comment by TYLEROL** [comment]
Hi @AngelRodriguez8008 - It looks like this issue #747 and #749 have conflicts due to merging of one of the earlier PRs.

Could you take a quick look and resolve those? 

Thanks,
Tyler

**Comment by AngelRodriguez8008** [comment]
Yes, I  implement these PRs independently, I will merge it now.  Thanks!

**Review by TYLEROL — APPROVED** [review]

**Comment by AngelRodriguez8008** [comment]
Hi @TYLEROL, 

I have a merge Issue hier. I add one required commit to this branch after you merge it, this is a Mayor Issue, Login was not working.

Now i am getting problem with the fix #746 (by @rswafford) OpenRecord is not working in some cases. I will try to fix it. 

Sorry.

BR

File: Microsoft.Dynamics365.UIAutomation.Browser/Extensions/SeleniumExtensions.cs
Comment (thathi) [inline Microsoft.Dynamics365.UIAutomation.Browser/Extensions/SeleniumExtensions.cs]: @AngelRodriguez8008 this is throwing StaleElementReferenceException before timeout? any Clue?

**Pull request merged**
